### DISCUSSION TITLE
Improvements of user experience with duplicates

### DISCUSSION THREAD

**Discussion by madbrain76** [music-assistant/support#2462] [Feature requests and ideas]
I have experienced several cases of issues with duplicates recently. 
1. False duplicates that caused some tracks to be hidden in the UI, and skipped during playback. This was arguably something caused by tagging not being sufficiently differentiated, but it was still difficult to figure out.
2. Duplicates not recognized as such, causing a track to be shown and played more than once.

I'm proposing several changes that are related, but even implementing just a subset of them would help.

1) The first case could be worked around by allowing duplicate detection to be turned off, either globally or within a specific music provider.

2) the first case could also be mitigated by enhancements to the matching algorithm, such as taking into account the track number metadata element, for the instance I ran into.

3) for instances of the first case not fixed by proposal 2), there should be a way the user to at least become aware that a given track or album has duplicates, hopefully before playback, other than reviewing the log

4) there could be some sort of duplicate management UI to facilitate matching/unmatching of tracks or albums. All duplifates woule be listed. Suspected duplicates could be listed too and sorted, depending on how many metadata elements differ for example.

Since this problem is usually resolved externally, perhaps display a message telling the user which program to run based on the provider, and which files/folders to retag/rename. Maybe link to some explanation page. Ideally, it would be possible to resolve within MA, but that seems difficult to design/implement, and would only work with some providers.

5) for the second case, if 2 tracks only differ by their favorite status, they should still be considered to be duplicates. Some UI change to the favorite display element would be needed in this case. I have some idea on this, such as showing a half heart or broken heart, pharmacy or red cross element - reallly, anytying to draw attention to the user. And, if clicked upon, show the 2 track locations with their distinct favorite heart, and let them be clicked to vary them individually.
Obviously there might be a better solution, this is just what came to mind..

**Comment by madbrain76** [comment]
As an example of sub-optimal experience, I loaded one album in 3 providers :

1. in the FS provider, tagged with Picard, as recommended, without any rating element
2. in the Plex provider, tagged by Plex , where it doesn't have any star rating
3. in the Qobuz provider, with their medata. The album is favorited, as is track 8
4. I then searched for two words in the title - "Grand Manner".

It found a single album, showing as favorite (heart). Presumably, because the album is favorite in Qobuz.
![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/6019f64c-e14b-4420-bf3a-2e9fdb7b11f4)

5. I clicked that album. I initially saw 36 tracks.

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/c6dd7c3a-9b40-41cc-b032-e57b10d086d7)

The first 18 tracks, which are not numbered, appear to come from the file system provider. They are all favorite (heart) even though the FS provider does not support favorites.
Underneath, the Plex version shows up, with tracks numbered from 1 to 18, but not favorited, as expected in this case.

6. I clicked the | menu next to one of the FS tracks and "show info". It showed this :

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/213ae77f-bec4-4a3a-967a-b2cde3925604)

A little odd that it has the FS provider icon( bookshelf) next to the album name under "appears on". But the provider details below say it's from "Plex Media Server Library".

7. I click the album again. Now, very strangely, I'm down to 35 tracks , from 36 earlier.

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/543845b8-f0b0-42e7-b8c4-9a571fa0201d)

8. I scrolled down and found that the Plex track 1 file is no longer listed.

It skips from the last FS entry to the first Plex entry with track 2, which is BWV 903 II Fugue.

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/37e97b19-80c4-49a0-8671-3f0a69e6b8f3)

Very odd.

9. Scrolling down all the way, the bottom shows :

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/0b32a4c1-d365-472f-8fe3-1d695e422bc2)

The Qobuz copy is listed under "Other versions".

10. If I click it, the list of tracks above does not change. But if I go all the way up, I see :

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/b7ee2378-89ab-4194-be3e-acc92d33f264)

Notice how there is no longer a track count next to "tracks". And the album name is prefixed with "Bach :". And there is a Qobuz provider icon.

11. Select a CCA player
12. click the blue Play button on top
13. Click the disc icon on the bottom left to open the player
14. click the queue icon on the lower right
15. click track 8
16. notice how the heart icon is empty, even though the track is favorited in Qobuz

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/282a1af7-7b24-4ef6-92d2-4c52bfb9763c)

There is also no indication on this screen that it's from Qobuz. Not sure if there should be.
17. using the "| Show info " menu on that track brings this :

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/db67dd14-6537-464d-8475-9454324448b5)

18. I pressed the back button in my browser. It brought me to this screen.
![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/14693274-d7bc-4a0e-8e65-f668512c1105)
There is a now track count of 19, vs 35 in step 7. All tracks are listed as belonging to the FS provider, and favorited, which is wrong.
The bottom of the screen shows this :

![image](https://github.com/music-assistant/hass-music-assistant/assets/57333073/f327f196-aa2e-4a22-b6d6-8a653faf9b81)

There were so many strange things that happened during this session that I feel compelled to include a log. I believe all the oddities I encountered should be reproducible with this test case, but one would need the FLAC files for the FS and Plex providers, and a Qobuz account, in order to do so. There are many little things happening that I'm not going to try to recap the issues - following the test case should be too hard. These are - showing 36 tracks (no match), showing mismatched tracks and providers, mismatched provider and favorite status, shrinking or disappearing total track counts, incomplete and changing music provider list for the album/tracks.
 
[log.txt](https://github.com/user-attachments/files/15846194/log.txt)

**Comment by marcelveldt** [comment]
Please limit your reports to short reports, straight to the point. I'm sure you mean well but this are complete books for me to read and not easy to quickly pickup from the pile of bugs to fix. 

Also be aware that this is still all in early stage so things are missing and we have super limited capacity. I do appreciate you taking the effort of making complete bug reports with reproduction steps but please keep them short(er) and leave out all irrelevant info so it's easier to digest. Often if I can read the issue report, reproduce it in limited time, the fix is often super fast.

Thanks!

**Comment by madbrain76** [comment]
@marcelveldt ,

I realize this one was a ton of stuff. I saw many different problems. I was not confident that I could each problem separately. 
It may well be that I can, but I haven't spent the time trying yet. It's also possible that the issues are sufficiently state-dependent that a large number of steps is still needed to surface them.

One alternative is that I forget about those problems altogether, and don't file anything. I don't think that's a great idea.
Or I could delay filing them until I have separated them, if I manage to. One problem is what to do before that potential separation. How do I store all the steps, screenshots and logs in order, in a way that I can separate things and work on them on a different date. I would need something like a "draft" issue. I don't know if that exists in github. Maybe it can be designated with a label, so it doesn't go immediately to triage ?

For most bugs, I try to rule out state problems, by first deleting and reinstalling MA. Of course, that increases the length of the test cases a bit, for the initial steps at least. I can remove those steps in the future if they are not relevant to triggering the bug. That will shorten some test cases. But for something like dupe issues, there is a lot of state involved with the multiple providers and metadata, and it's inevitable that the test cases will be longer than average.

Sorry to have written another book. This probably belongs outside of this particular RFE, but I don't know a better place.

**Comment by OzGav** [comment]
> They are all favorite (heart) even though the FS provider does not support favorites.

You still are not understanding what the heart is. It has NOTHING to do with the file system provider. It is a subset of the LIBRARY. 

> A little odd that it has the FS provider icon( bookshelf) next to the album name under "appears on". But the provider details below say it's from "Plex Media Server Library".

The little book icon is showing you that the track is in the library not that it has come from the file system provider. 

This report is too large and has comments that are not accurate so it is not possible to follow. There may be an edge case here but I can't be sure. I say may as what you are doing is not what anyone else would do and that is load up an album in Plex and the local provider. It makes no sense to do that however yes MA should probably be able to cope with it. Maybe it is and the linking is working correctly.

I request you review this report in light of the above information. It probably needs closing and another raised if there is indeed a linking problem.

**Comment by madbrain76** [comment]
@marcelveldt ,

Thanks, I didn't realize that's what the bookshelf meant. It invalidates many of the statements I made, but several odd behaviors remain, in terms of failed linking, track count showing different values and even disappearing.

II would urge you not to close this, at least because the OP makes clear feature requests (hopefully).

The comment should probably have been a separate issue. I was mainly trying to show the UX. Maybe I can move it to a new place, but I'm not sure if it should be a discussion or issue.

While no one may do this intentionally with 3 providers, problems could still happen unintentionally with 2 providers, FS + Qobuz, or FS + another music service. I don't think that case is uncommon. Some problems can still occur even with 2, and may not require as long of a test case.

**Reply by OzGav** [reply]
Firstly we really appreciate the effort you are going to with these reports. However, they are too big to follow when they contain multiple reported problems. Each one can only have one problem, for example, failed linking. What you can do is grab the second post and copy and paste it into an issue report focussing on one thing and keep it short and to the point removing any incorrect statements given the info above. Marcel wants to fix any linking problems as it is a core component of the software so again really appreciate you finding these.

**Reply by madbrain76** [reply]
@OzGav ,
I'm glad my efforts are appreciated.
I don't wish to write a very long tirade to justify writing large test cases with multiple issues such as this one. And yet I'm compelled to do so, even though I think it's the wrong forum.

As a developer, I also prefer single issue 100% reproducible bugs with short test cases. You'll get zero argument from me that it is something to strive for. I also prefer to review 1-file single line patches, vs something that affects 10 files in 3 different subsystem.

Sometimes, issues really do require long test cases to reproduce. And some bugs cannot be fixed with a set of consecutive small isolated patches without breaking the build.

In some cases, both test cases and patches can easily be reduced . In others, not, for issues that are complex, interdependent, or elusive.

What is certain is that working on reduction takes extra time. We are all volunteers, and it is a finite resource.

A long, detailed and reproducible test case is IMO still significantly preferable to not reporting the issues, or reporting the issue without any test case. Or reporting only the first issue. I had seen the varying track count before yesterday, for example, but had no clear test case for it, and didn't open an issue before for that reason. I waited until I had a test case to do so, even though it was a long one.

The reduction attempt can be performed either by the reporter, developer, or nobody, depending on resources.  But there needs to be workable workflow to go from large to small. It can happen on different days. I don't see an obvious solutIon here. You can't save your issue in progress in your favorite editor/IDE and then easily submit in github, like you can for code  There is no PR for issues, as far as I know.

I have actually tried to do this using an IDE/editor. There are problems pasting into github where step numbers disappear and need to be re-entered. I also can't easily attach screenshots and logs in the middle of the document prior to submitting. Even if I could save it locally in a usable format, it's not clear how I could submit it subsequently without a lot of busywork.

Being able to save the large test case info into giithub as a draft, without immediately submitting it, and prior to the reduction attempt, would go a long way. I'm open to suggestions. Simply being able to mark an issue as in progress would be enough. Similar to my previous request to let the reporter give a priority hint. I don't know how hard it is to add those features to the  github issue tracker.

TLDR test case reduction is highly desirable, but takes time, isn't always possible, and is not facilitated by the current github issue tracker.

**Comment by madbrain76** [comment]
Thinking about this some more, git PRs for test case reduction is probably not the best way to go :)

Things that would help :

1. allow an issue to be created from a comment containing a long test case - moving or duplicating that comment along with all attachments/screenshot . It looks like you can create an issue with a link to a comment, but there may not be a way to do a comment->issue copy or move. Search engines do show some demand for this feature.
ChatGPT came up with a Python program it claims can do it. It would need some permissions. It really should be in the web UI, but running a command-line is still better than rewriting the issue.
2. move other related comments from the old to the new issue
3. allow cloning an issue containing a large test case, so that it can be reduced in the copy/copies . I don't know if this feature exists. Google searches tells me that I'm far from alone in asking for it.
Again, ChatGPT claims it's possible in Python, with permissions.
4. allow an issue with a large test case to be saved as draft, or some other designation, so that it doesn't immediately go to triage, similar to the enhancement request to give a priority hint. Hopefully, this can be achieved with tagging

I think most if not all issue management systems I have worked with over the last 30 years allow for 3 and 4 in their UI. I can't recall if any supported the first two requests, through either UI or API. I might just not have been aware of it. None of these tools used markdown, or supported inline images or file attachments within the text.

The main problem with these Python programs is that they need permissions, and presumably only the repository owner could run them. But that is very unlikely to be the person doing the test case reduction. It would be the original bug filer, developer, or some other volunteer who would need it. Maybe the scripts and server-side could be configured to allow it if it's one of the first two. Or only for certain users.

**Comment by OzGav** [comment]
In 2.7.0 beta 24 and later there is now functionality to create and remove mappings. This is available in the PROVIDER DETAILS and OTHER VERSIONS sections of the item views
